### PR TITLE
ComponentOptions.buildJsonObjectOption<T>

### PR DESCRIPTION
**Pull request by francoislg**
Allows to configure `data-ranges=[{"start":100, "end":200, "label":"somewhere"}]` directly in the markup.

Comments are not only welcome but appreciated 😸 

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/Base/ComponentOptions.ts [hunk 1]
```diff
@@ -629,6 +636,19 @@ export class ComponentOptions {
...
 
+  static loadJsonObjectOption<T>(element: HTMLElement, name: string, option: IComponentOptions<any>): T {
+    const jsonAsString = ComponentOptions.loadStringOption(element, name, option);
+    if (jsonAsString == null) {
+      return null;
+    }
+    try {
+      return JSON.parse(jsonAsString) as T;
+    } catch (exception) {
+      new Logger(element).info(`Value for option ${name} is not a valid JSON string (Value is ${jsonAsString}). It has been disabled.`);
```
Comment (olamothe) [inline src/ui/Base/ComponentOptions.ts]: We could output the `exception` here also.

File: src/ui/Facet/Facet.ts [hunk 2]
```diff
@@ -608,9 +608,7 @@ export class Facet extends Component {
      * // });
      * ```
      */
-    valueCaption: ComponentOptions.buildCustomOption<IStringMap<string>>(() => {
-      return null;
-    }),
+    valueCaption: ComponentOptions.buildJsonObjectOption<IStringMap<string>>(),
```
Comment (olamothe) [inline src/ui/Facet/Facet.ts]: The documentation that accompany this option is now incorrect.

File: src/ui/FacetRange/FacetRange.ts [hunk 3]
```diff
@@ -120,9 +120,7 @@ export class FacetRange extends Facet implements IComponentBindings {
      * function (see [Query Function](https://developers.coveo.com/x/XQCq)). When this is the case, you must specify the
      * ranges at query time.
      */
-    ranges: ComponentOptions.buildCustomOption<IRangeValue[]>(() => {
-      return null;
-    }),
+    ranges: ComponentOptions.buildJsonObjectOption<IRangeValue[]>(),
```
Comment (olamothe) [inline src/ui/FacetRange/FacetRange.ts]: Same.

File: test/ui/ComponentOptionsTest.ts [hunk 4]
```diff
@@ -406,6 +413,27 @@ export function ComponentOptionsTest() {
...
+
+        it('which loads a JSON object option', () => {
+          const option = ComponentOptions.loadJsonObjectOption<IJsonObjectTest>(elem, 'myTestJsonObject', {});
+          expect(option.foo).not.toBeUndefined();
+          expect(option.foo).toBe('bar');
+        });
+        it('which loads a JSON array option', () => {
+          const option = ComponentOptions.loadJsonObjectOption<IJsonObjectTest[]>(elem, 'myTestJsonArray', {});
+          expect(option.length).toBe(1);
+          expect(option[0].foo);
```
Comment (dms1lva) [inline test/ui/ComponentOptionsTest.ts]: Does that do anything? Maybe you are looking for `toBeDefined`

**Comment by francoislg** [comment]
2017-05-31 11_11_55-facet _ coveo javascript search framework - reference documentation

Thoughts? I have done something similar with `FacetRange.ranges`.

I don't think we could update the typedoc generator to support this though, since it will always require a custom `T` type.

Also, pinging @fbeaudoincoveo for the doc :)

**Pull request merged**